### PR TITLE
How to create projects that target earlier frameworks

### LINKED ISSUE #24544: "Target frameworks in SDK-style projects" appears out of date for older versions of .NET Core
With .NET 5.0 installed it doesn't appear I can use target frameworks older than `netcoreapp3.1`. I get an invalid parameter error when trying any of the older version listed as options.

```bash
$ dotnet new webapi -f netcoreapp3.0
Couldn't find an installed template that matches the input, searching online for one that does...
Error: Invalid parameter(s):
-f netcoreapp3.0
    'netcoreapp3.0' is not a valid value for -f (Framework).
```

`net5.0` and `netcoreapp3.1` work, but I get the same error for all of these:

```
netcoreapp1.0
netcoreapp1.1
netcoreapp2.0
netcoreapp2.1
netcoreapp2.2
netcoreapp3.0
```

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 49205137-dcec-a120-e5e7-1b9c79cdafbb
* Version Independent ID: 4352856b-5ccf-8e62-48b4-3151340cac56
* Content: [Target frameworks in SDK-style projects - .NET](https://docs.microsoft.com/en-us/dotnet/standard/frameworks)
* Content Source: [docs/standard/frameworks.md](https://github.com/dotnet/docs/blob/main/docs/standard/frameworks.md)
* Product: **dotnet**
* Technology: **dotnet-standard**
* GitHub Login: @gewarren
* Microsoft Alias: **gewarren**

### PR DESCRIPTION
**Pull request by tdykstra**
Fixes #24544

File Type | File Name | Published Url
-- | -- | --
Content | docs/core/tools/dotnet-new-sdk-templates.md | https://review.docs.microsoft.com/en-us/dotnet/core/tools/dotnet-new-sdk-templates?branch=pr-en-us-25603

### CODE REVIEW

File: docs/core/tools/dotnet-new-sdk-templates.md [hunk 1]
```diff
@@ -88,6 +92,8 @@ Each template may have additional options available. The core templates have the
 
   Specifies the [framework](../../standard/frameworks.md) to target. Values: `net5.0` or `netcoreapp<version>` to create a .NET Class Library or `netstandard<version>` to create a .NET Standard Class Library. The default value for .NET 5.0 SDK is `net5.0`.
 
+  To create a project that targets a framework earlier than the SDK that you're using, see [`--framework` for `console` projects](#template-options) earlier in this article.
+
 - **`--langVersion <VERSION_NUMBER>`**
 
   Sets the `LangVersion` property in the created project file. For example, use `--langVersion 7.3` to use C# 7.3. Not supported for F#. Available since .NET Core 2.2 SDK.
```
Comment (BillWagner) [inline docs/core/tools/dotnet-new-sdk-templates.md:99]: Not related to your PR, but we should make an explicit statement about VB support for --langVersion.

**Review by BillWagner — APPROVED** [review]
This LGTM @tdykstra 

I had one question that we should address if it's a quick change.

**Comment by vlada-shubina** [comment]
@tdykstra could we please take a pause on this PR to discuss it?

It seems that different templates owners have different rules:
- common project templates and web templates are following package per tfm approach and the packages are available on NuGet.org; so installing older package from NuGet.org will work. 
- test, wpf and winforms templates are following single package approach. For WPF / winforms only supported TFMs are available (3.1, 5.0, 6.0) and for test templates only 6.0 is available. Additionally those packages are not available on NuGet.org. For those templates only installing older EOL SDK will probably help to get older framework available.
- additionally some templates support hidden --target-framework-override parameter which allows setting any text value as tfm. However this value is not validated, and will not be taken into account when evaluating language feature support, so using it after https://github.com/dotnet/templating/issues/3359 will probably result in the template that cannot be compiled.

Adding @KathleenDollard and @DamianEdwards here for discussion.

**Comment by DamianEdwards** [comment]
Ideally all templates should be updating to use the template identifier to produce a package per version and accept the framework version as an option, along with publishing all versions to NuGet.org. As I understand it, that way the templating engine can discover and suggest the correct package when a framework option value is passed that isn't supported by any installed templates, but **is** supported by a template in a package on NuGet.org. This is how the web templates work.

**Comment by KathleenDollard** [comment]
I agree that moving to a single approach (package per TFM, available on NuGet) would be extremely helpful. I think for Microsoft templates we just move this direction as fast as the teams can manage it. 

This seems overly complex for third party template authors. I think if templates are delivered to NuGet separately, the single package approach makes sense for these templates. Adding that so template authors happening upon this aren't confused and to clarify with Tom. 

@vlada-shubina when we complete that approach, then will @tdykstra text be correct?

**Comment by vlada-shubina** [comment]
imo, both single package and package per TFM approaches are possible and have their pros and cons; it's more about finding a common ground here. 
In case we want to continue with package per TFM test-templates, wpf and winforms should align their templates and publishing accordingly. 
In case we want to continue with single package, probably all repos and possibly template engine itself have some work to do. 
Single package seems easier approach for understanding, however however we probably should somehow distinguish supported and EOL tfms when displaying them to users and this approach might involve higher maintenance costs (for example in case hot fix is needed for 3.1 template, it should be applied to 3.1 and higher template pack).

> As I understand it, that way the templating engine can discover and suggest the correct package when a framework option value is passed that isn't supported by any installed templates, but **is** supported by a template in a package on NuGet.org.

This is not a case anymore, we removed implicit templates search on template instantiation to avoid unnecessary delays on searching. Potentially we can enable behavior when the user can do the search to check for package availability: `dotnet new webapi --search --framework netcoreapp3.0`, but we need to put the SDK templates back to search cache file (now that SDK templates are not checked for updates any longer any way it should be ok to have them in search cache). I can work on this if needed.

> This seems overly complex for third party template authors. I think if templates are delivered to NuGet separately, the single package approach makes sense for these templates. Adding that so template authors happening upon this aren't confused and to clarify with Tom.

Definitely agree, we should recommend creating single package for template authors in all tutorials. However should also probably give an explanation of advanced scenario too (template grouping was highlighted as understanding pain point in couple issues already).

>@vlada-shubina when we complete that approach, then will @tdykstra text be correct?

Tom's text is correct. 
I would probably only suggest to reorder paragraphs and put installing the SDK as the first option, and installing NuGet package as alternative, smth like:


>The ability to create a project for an earlier TFM depends on having that version of the SDK installed. For example, if you have only SDK 5.0 installed, then the only value available for `--framework` is `net5.0`. If you install SDK 3.1, the value `netcoreapp3.1` becomes available for `--framework`. If you install SDK 2.1, `netcoreapp2.1` becomes available, and so on. So by specifying `--framework netcoreapp2.1` you can use SDK 2.1 even while running `dotnet new` in SDK 5.0.

>Alternatively, to create a project that targets a framework earlier than the SDK that you're using, you might need to install the NuGet package for the template. [Common](https://www.nuget.org/packages?q=Microsoft.DotNet.Common.ProjectTemplates), [web](https://www.nuget.org/packages?q=Microsoft.DotNet.Web.ProjectTemplates), and [SPA](https://www.nuget.org/packages?q=Microsoft.DotNet.Web.Spa.ProjectTemplates) project types use different packages per target framework moniker (TFM). For example, to be able to create a `console` project that targets `netcoreapp1.0`, run [`dotnet new --install`](dotnet-new-install.md) on `Microsoft.DotNet.Common.ProjectTemplates.1.x`.

**Comment by tdykstra** [comment]
> Tom's text is correct.

With your reordering of the options applied, is this correct for the current state of things, meaning we could merge this PR and revise the text when we make the changes under discussion?  Or hold the PR open until then?

**Review by vlada-shubina — APPROVED** [review]

**Comment by vlada-shubina** [comment]
Fine with me. I will open the tickets to test-templates, wpf and winforms repos (those that are not available on NuGet.org and following different packaging approach) tomorrow to continue discussion.

**Pull request merged**